pkg/controller/seed-controller-manager/initialmachinedeployment/controller.go

```diff
@@ -0,0 +1,243 @@
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		r.ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionMachineDeploymentControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(cluster)
+		},
+	)
+	if err != nil {
+		r.log.Errorw("Failed to reconcile cluster", "cluster", cluster.Name, zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// If cluster is not healthy yet there is nothing to do.
+	// If it gets healthy we'll get notified by the event. No need to requeue.
+	if !cluster.Status.ExtendedHealth.AllHealthy() {
+		return nil, nil
+	}
+
+	request := cluster.Annotations[v1.InitialMachineDeploymentRequestAnnotation]
+
+	// if there is no (more) annotation, we're done
+	if request == "" {
+		return nil, nil
+	}
+
+	var nodeDeployment *v1.NodeDeployment
+	if err := json.Unmarshal([]byte(request), &nodeDeployment); err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("cannot unmarshal initial MachineDeployment request: %v", err)
+	}
+
+	nodeDeployment, err := machineresource.Validate(nodeDeployment, cluster.Spec.Version.Semver())
+	if err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("initial node deployment is not valid: %v", err)
+	}
+
+	userClusterClient, err := r.userClusterConnectionProvider.GetClient(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user cluster client: %v", err)
+	}
+
+	if err := r.createInitialMachineDeployment(nodeDeployment, cluster, userClusterClient); err != nil {
+		return nil, fmt.Errorf("failed to create initial MachineDeployment: %v", err)
+	}
+
+	if err := r.removeAnnotation(cluster); err != nil {
+		return nil, fmt.Errorf("failed to remove initial MachineDeployment annotation: %v", err)
+	}
+
+	return nil, nil
+}
+
+func (r *Reconciler) createInitialMachineDeployment(nodeDeployment *v1.NodeDeployment, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client) error {
+	datacenter, err := r.getTargetDatacenter(cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get target datacenter: %v", err)
+	}
+
+	sshKeys, err := r.getSSHKeys(cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get SSH keys: %v", err)
+	}
+
+	data := common.CredentialsData{
+		Ctx:               r.ctx,
+		KubermaticCluster: cluster,
+		Client:            r,
+	}
+
+	machineDeployment, err := machineresource.Deployment(cluster, nodeDeployment, datacenter, sshKeys, data)
+	if err != nil {
+		return fmt.Errorf("failed to assemble MachineDeployment: %v", err)
+	}
+
+	err = client.Create(r.ctx, machineDeployment)
```

Comment:
What do you think about adding some event on success too?

Reply:
Added an event.